core/src/main/java/hudson/security/BasicAuthenticationFilter.java

```diff
@@ -138,7 +141,12 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
             User u = User.getById(username, true);
             ApiTokenProperty t = u.getProperty(ApiTokenProperty.class);
             if (t!=null && t.matchesPassword(password)) {
-                SecurityContextHolder.getContext().setAuthentication(u.impersonate());
+                UserDetails userDetails = u.getUserDetailsForImpersonation();
+                Authentication auth = u.impersonate(userDetails);
+
+                SecurityListener.fireAuthenticated(userDetails);
```

Comment:
I think that at this point userDetails cannot be null

https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/jenkins/security/SecurityListener.java#L83-L96

Reply:
I agree with @kuisathaverat . The null value must be handled somehow here

Reply:
Corrected by using an inner class to manage the UserDetails in such case.